### PR TITLE
[AD-227] Added support for CURRENT_TIMESTAMP, CURRENT_DATE, QUARTER, DAYNAME, and MONTHNAME.

### PR DESCRIPTION
**Pull request by birschick-bq**
### Summary

[AD-227] Added support for CURRENT_TIMESTAMP, CURRENT_DATE, QUARTER, DAYNAME, and MONTHNAME.

### Description

- CURRENT_TIMESTAMP/NOW, 
- CURRENT_DATE/TODAY, 
- QUARTER, 
- DAYNAME, and 
- MONTHNAME.
- Date/Datetime operators `+` and `-` (in Tableau)
- Small improvement in handling BsonTimestamp conversions.

### Related Issue

https://bitquill.atlassian.net/browse/AD-227

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java [hunk 1]
```diff
@@ -188,6 +204,8 @@ protected static String getNormalizedIdentifier(final String fieldName) {
 
         private static final Map<SqlOperator, String> MONGO_OPERATORS =
                 new HashMap<>();
+        private static final Map<SqlOperator,
+                BiFunction<RexCall, List<String>, String>> REX_CALL_TO_MONGO_MAP = new HashMap<>();
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java:201]: 💯

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java [hunk 2]
```diff
@@ -497,40 +604,434 @@ void testQueryExtract() throws SQLException {
...
+                    resultSet.getString(1));
+
+            Assertions.assertFalse(resultSet.next());
+        } finally {
+            Locale.setDefault(originalLocale);
+        }
+    }
+
+    /**
+     * Tests that CURRENT_TIMESTAMP.
```
Comment (andiemontoyeah) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java]: Tests that CURRENT_TIMESTAMP -> Tests CURRENT_TIMESTAMP?

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java [hunk 3]
```diff
@@ -497,40 +604,434 @@ void testQueryExtract() throws SQLException {
...
+        final ResultSet resultSet4 = statement.executeQuery(
+                String.format("SELECT \"field\" "
+                        + " FROM \"%s\".\"%s\""
+                        + " WHERE \"field\" > CURRENT_TIMESTAMP", DATABASE_NAME, tableName));
+        Assertions.assertNotNull(resultSet4);
+        Assertions.assertFalse(resultSet4.next());
+    }
+
+    /**
+     * Tests that CURRENT_DATE.
```
Comment (andiemontoyeah) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java]: Tests CURRENT_DATE?

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java [hunk 4]
```diff
@@ -497,40 +604,434 @@ void testQueryExtract() throws SQLException {
...
+        final ResultSet resultSet4 = statement.executeQuery(
+                String.format("SELECT \"field\" "
+                        + " FROM \"%s\".\"%s\""
+                        + " WHERE \"field\" > CURRENT_DATE", DATABASE_NAME, tableName));
+        Assertions.assertNotNull(resultSet4);
+        Assertions.assertFalse(resultSet4.next());
+    }
+
+    /**
+     * Tests that CURRENT_TIME.
```
Comment (andiemontoyeah) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementBasicTest.java]: Tests CURRENT_TIME?

**Review by andiemontoyeah — COMMENTED** [review]
lgtm!

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java [hunk 5]
```diff
@@ -100,12 +102,25 @@ public DocumentDbMqlQueryContext get(final String sql) throws SQLException {
             }
         } catch (Exception e) {
             throw SqlError.createSQLException(
-                    LOGGER, SqlState.INVALID_QUERY_EXPRESSION, e, SqlError.SQL_PARSE_ERROR, sql);
+                    LOGGER, SqlState.INVALID_QUERY_EXPRESSION, e, SqlError.SQL_PARSE_ERROR, sql,
+                    getExceptionMessages(e));
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java:106]: Recursively, gets the causing message and any suppressed messages.

File: docker/TableauConnector/dialect.tdd [hunk 6]
```diff
@@ -497,21 +496,36 @@
...
+    </function>
     <date-function name='DATEADD' return-type='datetime'>
       <formula>TIMESTAMPADD(%1, %2, %3)</formula>
       <argument type='localstr' />
       <argument type='int' />
       <argument type='datetime' />
     </date-function>
+    <date-function name='DATEDIFF' return-type='int'>
+      <formula>TIMESTAMPDIFF(%1, %2, %3)</formula>
+      <formula part='year'>CAST( EXTRACT(YEAR FROM CAST(%3 AS TIMESTAMP)) - EXTRACT(YEAR FROM CAST(%2 AS TIMESTAMP)) AS BIGINT)</formula>
```
Comment (andiemontoyeah) [inline docker/TableauConnector/dialect.tdd:518]: postgresql does not have support for any kind of datediff/ timestampdiff and mapped all datediff formulas to be some combination of EXTRACT and math. Copying their formula for year, month, quarter worked.

**Review by andiemontoyeah — APPROVED** [review]

**Review by mitchell-elholm — APPROVED** [review]
LGTM

**Review by andiemontoyeah — APPROVED** [review]

**Pull request merged**